### DISCUSSION TITLE
classes/class_tilemaplayer

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#27] [User-contributed notes]
# classes/class_tilemaplayer

Inherits: Node2D< CanvasItem< Node< Object Node for 2D tile-based maps. Description: Node for 2D tile-based maps. A TileMapLayer uses a TileSet which contain a list of tiles which are used to creat...

https://docs.godotengine.org/en/latest/classes/class_tilemaplayer.html

**Comment by boholstapp** [comment]
I tested this and it works https://www.youtube.com/watch?v=7_fwNiZL0xY&t=59s

**Comment by benmillane** [comment]
Is there a way to have TileMapLayers which use the same Tileset to share modulation settings? I can't seem to figure it out, and I'm thinking it might be better to just make a tilemaplayer its own scene, and modulate that instead?

**Reply by Leotin16** [reply]
I'm no entirely shure (and i'm certainly late), but can't you put some tileset layers into a plain 2D Node and modulate that 2D Node?

**Comment by NancokPS2** [comment]
The tutorial should probably be linked here.
https://docs.godotengine.org/en/stable/tutorials/2d/using_tilemaps.html

**Reply by Calinou** [reply]
Pull request opened:

- https://github.com/godotengine/godot/pull/101299

**Comment by troynall** [comment]
Centralized Modulate Controller

Create a small helper script to apply a color to all TileMapLayer children.

```gdscript
func set_modulate_all_layers(color: Color):
    for child in $TileMap.get_children():
        if child is TileMapLayer:
            child.modulate = color
```

Call this when you want to tint everything. Useful if your layers are already organized

**Comment by PangolinMontanari** [comment]
It seems like tilemaps with scene collections don't populate themselves with the scene tiles until after the scene is ready- even the cell data doesn't reflect the scene tiles immediately after instantiation(unlike the basic atlas tiles). This can cause some shenanigans if you want to open a tilemaplayer and then read its data immediately. Best bet is to await get_tree().process_frame.

**Reply by nielmclaren** [reply]
Encountered this as well. In my case I needed to `await get_tree().process_frame` twice.

A couple issues containing further workarounds / advice:
https://github.com/godotengine/godot/issues/67679
https://github.com/godotengine/godot/issues/76349

Hope this helps someone find the underlying problem a little sooner. (-:

**Comment by HeraldOfFire270** [comment]
Could probably use a small addendum to <code>get_used_rect()</code>.  The wording suggests the Recti returned would encompass the entire TileMapLayer, but it returns the number of tiles in each dimension without consideration for their size.

**Reply by NancokPS2** [reply]
It is kind of implied due to the return type.  
There would be no reason to use a Rect2i for a regular size.
